akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -194,99 +214,89 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
         }
         def connectionWindowAvailable(): MultiplexerState = this // nothing to do, as there is no data to send
         def enqueueOutStream(streamId: Int): MultiplexerState =
-          if (connectionWindowLeft == 0) WaitingForConnectionWindow(immutable.TreeSet(streamId))
-          else sendDataFrame(streamId, Set.empty)
+          if (connectionWindowLeft == 0) {
+            sendableOutstreams += streamId
+            WaitingForConnectionWindow
+          } else sendDataFrame(streamId)
         def closeStream(streamId: Int): MultiplexerState = this
       }
 
       /** Not yet pulled but data waiting to be sent */
-      private[http2] case class WaitingForNetworkToSendControlFrames(controlFrameBuffer: immutable.Vector[FrameEvent], sendableOutstreams: immutable.Set[Int]) extends MultiplexerState {
-        require(controlFrameBuffer.nonEmpty)
-        allowReadingIncomingFrames(controlFrameBuffer.size < settings.outgoingControlFrameBufferSize)
-        def onPull(): MultiplexerState = controlFrameBuffer match {
-          case first +: remaining =>
-            pushFrameOut(first)
-            allowReadingIncomingFrames(remaining.length < settings.outgoingControlFrameBufferSize)
-            if (remaining.isEmpty && sendableOutstreams.isEmpty) Idle
-            else if (remaining.isEmpty) WaitingForNetworkToSendData(sendableOutstreams)
-            else copy(remaining, sendableOutstreams)
+      case object WaitingForNetworkToSendControlFrames extends MultiplexerState {
+        def onPull(): MultiplexerState = {
+          val first = controlFrameBuffer.dequeue()
+          pushFrameOut(first)
+          if (controlFrameBuffer.isEmpty && sendableOutstreams.isEmpty) Idle
+          else if (controlFrameBuffer.isEmpty) WaitingForNetworkToSendData
+          else this
+        }
+        def pushControlFrame(frame: FrameEvent): MultiplexerState = {
+          controlFrameBuffer += frame
+          this
         }
-        def pushControlFrame(frame: FrameEvent): MultiplexerState = copy(controlFrameBuffer = controlFrameBuffer :+ frame)
         def connectionWindowAvailable(): MultiplexerState = this
-        def enqueueOutStream(streamId: Int): MultiplexerState =
-          if (!sendableOutstreams.contains(streamId))
-            copy(sendableOutstreams = sendableOutstreams + streamId)
-          else
-            this
+        def enqueueOutStream(streamId: Int): MultiplexerState = {
+          sendableOutstreams += streamId
+          this
+        }
 
-        def closeStream(streamId: Int): MultiplexerState =
-          if (sendableOutstreams.contains(streamId)) {
-            val sendableExceptClosed = sendableOutstreams - streamId
-            copy(sendableOutstreams = sendableExceptClosed)
-          } else
-            this
+        def closeStream(streamId: Int): MultiplexerState = {
+          // leave stream in sendableOutstreams, to be skipped in sendNextFrame
```

Comment:
I guess you mean in `sendDataFrame`?
```suggestion
          // leave stream in sendableOutstreams, to be skipped in sendDataFrame
```

Reply:
Indeed, thanks.